src/unistd.rs

```diff
@@ -2392,3 +2397,425 @@ pub fn access<P: ?Sized + NixPath>(path: &P, amode: AccessFlags) -> Result<()> {
     })?;
     Errno::result(res).map(drop)
 }
+
+/// Representation of a User, based on `libc::passwd`
+///
+/// The reason some fields in this struct are `String` and others are `CString` is because some
+/// fields are based on the user's locale, which could be non-UTF8, while other fields are
+/// guaranteed to conform to [NAME_REGEX](https://serverfault.com/a/73101/407341), which only
+/// contains ASCII.
+#[derive(Debug, Clone, PartialEq, Default)]
+pub struct User {
+    /// Username
+    pub name: String,
+    /// User password (probably encrypted)
+    pub passwd: CString,
+    /// User ID
+    pub uid: Uid,
+    /// Group ID
+    pub gid: Gid,
+    /// User information
+    #[cfg(not(target_os = "android"))]
+    pub gecos: CString,
+    /// Home directory
+    pub dir: PathBuf,
+    /// Path to shell
+    pub shell: PathBuf,
+    /// Login class
+    #[cfg(not(any(target_os = "android", target_os = "linux")))]
+    pub class: CString,
+    /// Last password change
+    #[cfg(not(any(target_os = "android", target_os = "linux")))]
+    pub change: libc::time_t,
+    /// Expiration time of account
+    #[cfg(not(any(target_os = "android", target_os = "linux")))]
+    pub expire: libc::time_t
+}
+
+impl From<*mut libc::passwd> for User {
```

Comment:
Does not work. 

I get:

```
error[E0277]: the trait bound `unistd::User: std::convert::From<*mut libc::passwd>` is not satisfied
    --> src/unistd.rs:2713:25
     |
2713 |                 Some(Ok(User::from(res)))
     |                         ^^^^^^^^^^ the trait `std::convert::From<*mut libc::passwd>` is not implemented for `unistd::User`
```

Reply:
Well, yeah.  You're going to have to convert the mutable pointer into a constant one.  But that's easy.  Going the other way isn't possible.  That's why it's preferable to implement `From` for the constant pointer type, not the mutable type.  In fact, I think that conversion would be automatic if you implemented `From` on the reference type instead of the pointer type.  That would look something like this:
```rust
impl From<&libc::passwd> for User {...}
...
if error == ERANGE {
    return some_kind_of_error;
} else if error != 0 {
    return some_other_kind_of_error;
} else {
    return res.as_ref().map(User::from);
}
```